### PR TITLE
[Reward] initial CLoud Reward trainer

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Adds an option to the `RewardTrainer` to implement the CLoud method.

### CODE REVIEW

File: docs/source/reward_trainer.mdx [hunk 1]
```diff
@@ -81,6 +81,15 @@ training_args = RewardConfig(
 
 For reference results, please refer PR [#1932](https://github.com/huggingface/trl/pull/1932).
 
+## CLoud Critique-out-Loud Reward Models
+
+[Critique-out-Loud reward models](https:/huggingface.co/papers/2408.11791) are reward models that can reason explicitly about the quality of an input through producing Chain-of-Thought like critiques of an input before predicting a reward. In classic reward model training, the reward model is trained as a reward head initialized on top of the base LLM. Without LM capabilities, classic reward models act as encoders and must predict rewards within a single forward pass through the model, meaning reasoning must happen implicitly. In contrast, CLoud reward models are trained to both produce explicit reasoning about quality and to score based on these critique reasoning traces. 
+
+To train a Critique-out-Loud reward model, you can use the `feedback_method="teacher"` and set the `lm_weight` to a high value.
```
Comment (qgallouedec) [inline docs/source/reward_trainer.mdx:88]: How to set the lm_height to a high value? Maybe add a ref or a code example

File: docs/source/reward_trainer.mdx [hunk 2]
```diff
@@ -81,6 +81,15 @@ training_args = RewardConfig(
...
+
+To train a Critique-out-Loud reward model, you can use the `feedback_method="teacher"` and set the `lm_weight` to a high value. 
+
+The dataset should contain the columns `"prompt"`, `"chosen"`, `"rejected"` as well as the `"chosen_feedback` and `"rejected_feedback"` columns which contain the Chain-of-Thought like critiques for the chosen and rejected responses respectively generated by the same model or a different model. A script to generate such a dataset from a dataset of preferences is available in the `examples/datasets/critique_out_loud_vllm.py` file.
```
Comment (qgallouedec) [inline docs/source/reward_trainer.mdx:90]: Add a link to the example

File: tests/test_reward_trainer.py [hunk 3]
```diff
@@ -240,3 +240,146 @@ def test_tags(self):
                 model=self.model, args=training_args, processing_class=self.tokenizer, train_dataset=dummy_dataset
             )
             self.assertEqual(trainer.model.model_tags, trainer._tag_names)
+
+    def test_train_with_feedback(self):
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            # Create a dummy dataset with feedback
+            dummy_dataset_dict = {
```
Comment (qgallouedec) [inline tests/test_reward_trainer.py:247]: Ok now I get why you want a tiny version: the zen dataset misses two columns.
I'll think about the best way to do it

**Comment by qgallouedec** [comment]
I'm closing this PR because no recent activity. Please feel free to reopen it or request a reopening if necessary.
